smart_card_connector_app/src/application_unittest.cc

```diff
@@ -183,33 +223,114 @@ TEST_F(SmartCardConnectorApplicationTest, InternalApiSingleDeviceListing) {
                                   /*pvReserved2=*/nullptr, &scard_context),
             SCARD_S_SUCCESS);
+  std::vector<std::string> readers = DirectCallSCardListReaders(scard_context);
+  EXPECT_EQ(SCardReleaseContext(scard_context), SCARD_S_SUCCESS);
 
-  DWORD readers_size = 0;
-  EXPECT_EQ(SCardListReaders(scard_context, /*mszGroups=*/nullptr,
-                             /*mszReaders=*/nullptr, &readers_size),
+  // Assert:
+
+  EXPECT_THAT(readers, ElementsAre("Gemalto PC Twin Reader 00 00"));
+}
+
+TEST_F(SmartCardConnectorApplicationTest,
+       InternalApiGetStatusChangeDeviceAppearing) {
+  // Arrange:
+
+  // Start with an empty list of readers.
+  StartApplication();
+  EXPECT_TRUE(reader_notification_observer().Empty());
+
+  // Act:
+
+  SCARDCONTEXT scard_context = 0;
+  EXPECT_EQ(SCardEstablishContext(SCARD_SCOPE_SYSTEM, /*pvReserved1=*/nullptr,
+                                  /*pvReserved2=*/nullptr, &scard_context),
             SCARD_S_SUCCESS);
-  EXPECT_GT(readers_size, 0U);
+  EXPECT_TRUE(DirectCallSCardListReaders(scard_context).empty());
 
-  std::string readers_multistring;
-  readers_multistring.resize(readers_size);
-  EXPECT_EQ(SCardListReaders(scard_context, /*mszGroups=*/nullptr,
-                             &readers_multistring[0], &readers_size),
+  // Simulate connecting a reader.
+  TestingSmartCardSimulation::Device device;
+  device.id = 123;
+  device.type = TestingSmartCardSimulation::DeviceType::kGemaltoPcTwinReader;
+  SetUsbDevices({device});
+
+  // Wait until PC/SC reports a change in the list of readers.
+  std::vector<SCARD_READERSTATE> reader_states(1);
+  reader_states[0].szReader = R"(\\?PnP?\Notification)";
+  reader_states[0].dwCurrentState = SCARD_STATE_UNAWARE;
+  EXPECT_EQ(SCardGetStatusChange(scard_context, /*dwTimeout=*/INFINITE,
+                                 reader_states.data(), reader_states.size()),
             SCARD_S_SUCCESS);
-  EXPECT_EQ(readers_size, readers_multistring.size());
+
+  std::vector<std::string> readers = DirectCallSCardListReaders(scard_context);
 
   EXPECT_EQ(SCardReleaseContext(scard_context), SCARD_S_SUCCESS);
 
   // Assert:
 
-  EXPECT_THAT(ExtractMultiStringElements(readers_multistring),
+  EXPECT_THAT(readers, ElementsAre("Gemalto PC Twin Reader 00 00"));
+  EXPECT_EQ(reader_notification_observer().WaitAndPop(),
+            "reader_init_add:Gemalto PC Twin Reader");
+  EXPECT_EQ(reader_notification_observer().WaitAndPop(),
+            "reader_finish_add:Gemalto PC Twin Reader");
+  EXPECT_TRUE(reader_notification_observer().Empty());
+}
+
+TEST_F(SmartCardConnectorApplicationTest,
+       InternalApiGetStatusChangeDeviceRemoving) {
+  // Arrange:
+
+  // Start with a single reader.
+  TestingSmartCardSimulation::Device device;
+  device.id = 123;
+  device.type = TestingSmartCardSimulation::DeviceType::kGemaltoPcTwinReader;
+  SetUsbDevices({device});
+
+  StartApplication();
+  EXPECT_EQ(reader_notification_observer().Pop(),
```

Comment:
I guess we have some guarantee that WaitAndPop is not required here

Reply:
Added a clarification comment.

(BTW - between us - I myself don't fully understand why we need to wait in the scenarios added in this PR ::confused: . When I didn't wait, it failed on Github's VMs, but only in Release mode and not reproducible for me locally. But I suppose it might be legitimate, because PC/SC's threading model is not simple.)